File: bindings/java/src/test/java/dev/kairoscode/zlink/integration/bench/BenchPair.java
package dev.kairoscode.zlink.integration.bench;

import dev.kairoscode.zlink.*;

import java.util.concurrent.atomic.AtomicBoolean;

final class BenchPair {
    private BenchPair() {
    }

    static int run(String transport, int size) {
        return runPairLike("PAIR", SocketType.PAIR, SocketType.PAIR, transport, size);
    }

    static int runPairLike(String outPattern, SocketType aType, SocketType bType,
                           String transport, int size) {
        int warmup = BenchUtil.parseEnv("BENCH_WARMUP_COUNT", 1000);
        int latCount = BenchUtil.parseEnv("BENCH_LAT_COUNT", 500);
        int msgCount = BenchUtil.resolveMsgCount(size);

        Context ctx = new Context();
        Socket a = new Socket(ctx, aType);
        Socket b = new Socket(ctx, bType);

        try {
            String endpoint = BenchUtil.endpointFor(transport, outPattern.toLowerCase());
            a.bind(endpoint);
            b.connect(endpoint);
            BenchUtil.sleep(300);

            byte[] buf = new byte[size];
            for (int i = 0; i < size; i++) {
                buf[i] = 'a';
            }

            for (int i = 0; i < warmup; i++) {
                b.send(buf, SendFlag.NONE);
                a.recv(size, ReceiveFlag.NONE);
            }

            long t0 = System.nanoTime();
            for (int i = 0; i < latCount; i++) {
                b.send(buf, SendFlag.NONE);
                byte[] x = a.recv(size, ReceiveFlag.NONE);
                a.send(x, SendFlag.NONE);
                b.recv(size, ReceiveFlag.NONE);
            }
            double latUs = (System.nanoTime() - t0) / 1000.0 / (latCount * 2.0);

            AtomicBoolean recvDone = new AtomicBoolean(false);
            AtomicBoolean recvFail = new AtomicBoolean(false);
            Thread receiver = new Thread(() -> {
                try {
                    for (int i = 0; i < msgCount; i++) {
                        a.recv(size, ReceiveFlag.NONE);
                    }
                    recvDone.set(true);
                } catch (Exception e) {
                    recvFail.set(true);
                }
            });

            receiver.start();
            t0 = System.nanoTime();
            for (int i = 0; i < msgCount; i++) {
                b.send(buf, SendFlag.NONE);
            }
            receiver.join();

            if (!recvDone.get() || recvFail.get()) {
                return 2;
            }

            double sec = (System.nanoTime() - t0) / 1_000_000_000.0;
            double thr = msgCount / sec;
            BenchUtil.printResult(outPattern, transport, size, thr, latUs);
            return 0;
        } catch (Exception e) {
            return 2;
        } finally {
            try {
                a.close();
            } catch (Exception ignored) {
            }
            try {
                b.close();
            } catch (Exception ignored) {
            }
            try {
                ctx.close();
            } catch (Exception ignored) {
            }
        }
    }
}

File: bindings/dotnet/perf/single/Zlink.BindingBench/BenchPair.cs
using System;
using System.Diagnostics;
using System.Threading;
using Zlink;

internal static partial class BenchRunner
{
    internal static int RunPair(string transport, int size) =>
        RunPairLike("PAIR", SocketType.Pair, SocketType.Pair, transport, size);

    internal static int RunDealerDealer(string transport, int size) =>
        RunPairLike("DEALER_DEALER", SocketType.Dealer, SocketType.Dealer, transport, size);

    private static int RunPairLike(string outPattern, SocketType aType, SocketType bType, string transport, int size)
    {
        int warmup = ParseEnv("BENCH_WARMUP_COUNT", 1000);
        int latCount = ParseEnv("BENCH_LAT_COUNT", 500);
        int msgCount = ResolveMsgCount(size);

        using var ctx = new Context();
        using var a = new Zlink.Socket(ctx, aType);
        using var b = new Zlink.Socket(ctx, bType);

        try
        {
            string ep = EndpointFor(transport, outPattern.ToLowerInvariant());
            a.Bind(ep);
            b.Connect(ep);
            Thread.Sleep(300);

            var buf = new byte[size];
            Array.Fill(buf, (byte)'a');
            var recv = new byte[size];

            for (int i = 0; i < warmup; i++)
            {
                SendRetry(b, buf.AsSpan(), SendFlags.None);
                ReceiveRetry(a, recv.AsSpan(), ReceiveFlags.None);
            }

            var sw = Stopwatch.StartNew();
            for (int i = 0; i < latCount; i++)
            {
                SendRetry(b, buf.AsSpan(), SendFlags.None);
                int n = ReceiveRetry(a, recv.AsSpan(), ReceiveFlags.None);
                if (n != size)
                    return 2;
                SendRetry(a, recv.AsSpan(0, n), SendFlags.None);
                ReceiveRetry(b, recv.AsSpan(), ReceiveFlags.None);
            }
            sw.Stop();
            double latUs = (sw.Elapsed.TotalMilliseconds * 1000.0) / (latCount * 2);

            var recvDone = new ManualResetEventSlim(false);
            Exception? recvError = null;
            var th = new Thread(() =>
            {
                var recvThr = new byte[size];
                try
                {
                    for (int i = 0; i < msgCount; i++)
                        ReceiveRetry(a, recvThr.AsSpan(), ReceiveFlags.None);
                    recvDone.Set();
                }
                catch (Exception ex)
                {
                    recvError = ex;
                }
            });

            th.Start();
            sw.Restart();
            for (int i = 0; i < msgCount; i++)
                SendRetry(b, buf.AsSpan(), SendFlags.None);
            th.Join();
            sw.Stop();

            if (recvError != null || !recvDone.IsSet)
                return 2;

            double thr = msgCount / sw.Elapsed.TotalSeconds;
            PrintResult(outPattern, transport, size, thr, latUs);
            return 0;
        }
        catch
        {
            return 2;
        }
    }
}
